graph/src/data/schema.rs

```diff
@@ -1,28 +1,61 @@
-use data::graphql::validation::validate_schema;
+use data::graphql::validation::{get_object_type_definitions, validate_schema};
 use data::subgraph::SubgraphId;
 use failure::Error;
 use graphql_parser;
-use graphql_parser::{schema, Pos};
+use graphql_parser::{
+    query,
+    schema::{self, ObjectType},
+    Pos,
+};
+use std::collections::BTreeMap;
 
 /// A validated and preprocessed GraphQL schema for a subgraph.
 #[derive(Clone, Debug, PartialEq)]
 pub struct Schema {
     pub id: SubgraphId,
     pub document: schema::Document,
+    // Maps an interface name to the list of entities that implement it.
+    types_for_interface: BTreeMap<query::Name, Vec<ObjectType>>,
 }
 
 impl Schema {
     pub fn parse(raw: &str, id: SubgraphId) -> Result<Self, Error> {
         let document = graphql_parser::parse_schema(&raw)?;
         validate_schema(&document)?;
+
+        let mut types_for_interface = BTreeMap::<_, Vec<_>>::new();
+        for object_type in get_object_type_definitions(&document) {
+            for implemented_interface in object_type.implements_interfaces.clone() {
+                types_for_interface
+                    .entry(implemented_interface)
+                    .and_modify(|vec| vec.push(object_type.clone()))
+                    .or_insert(vec![object_type.clone()]);
+            }
+        }
+
         let mut schema = Schema {
             id: id.clone(),
             document,
+            types_for_interface,
         };
         schema.add_subgraph_id_directives(id);
+
         Ok(schema)
     }
 
+    pub fn type_for_interface(&self, interface_name: &query::Name) -> Option<&Vec<ObjectType>> {
+        self.types_for_interface.get(interface_name)
+    }
+
+    // #[cfg(test)]
```

Comment:
Should we remove this is if it's commented out anyway?

Reply:
Ouch forgot that.